include/knowhere/comp/knowhere_check.h

```diff
@@ -38,6 +42,93 @@ SuppportMmapIndexTypeCheck(const std::string& index_name) {
+    static std::set<knowhere::IndexType> has_raw_data_index_set = {
+        IndexEnum::INDEX_FAISS_BIN_IDMAP,  IndexEnum::INDEX_FAISS_BIN_IVFFLAT, IndexEnum::INDEX_FAISS_IVFFLAT,
+        IndexEnum::INDEX_FAISS_IVFFLAT_CC, IndexEnum::INDEX_HNSW_SQ8_REFINE,   IndexEnum::INDEX_SPARSE_INVERTED_INDEX,
+        IndexEnum::INDEX_SPARSE_WAND};
+    static std::set<knowhere::IndexType> has_raw_data_index_alias_set = {"IVFBIN", "BINFLAT", "IVFFLAT", "IVFFLATCC"};
+
+    static std::set<knowhere::IndexType> no_raw_data_index_set = {
+        IndexEnum::INDEX_FAISS_IVFPQ,     IndexEnum::INDEX_FAISS_IVFSQ8,      IndexEnum::INDEX_HNSW_SQ8,
+        IndexEnum::INDEX_FAISS_GPU_IDMAP, IndexEnum::INDEX_FAISS_GPU_IVFFLAT, IndexEnum::INDEX_FAISS_GPU_IVFSQ8,
+        IndexEnum::INDEX_FAISS_GPU_IVFPQ, IndexEnum::INDEX_RAFT_CAGRA,        IndexEnum::INDEX_GPU_CAGRA,
+        IndexEnum::INDEX_RAFT_IVFFLAT,    IndexEnum::INDEX_GPU_IVFFLAT,       IndexEnum::INDEX_RAFT_IVFPQ,
+        IndexEnum::INDEX_GPU_IVFPQ,
+    };
+
+    static std::set<knowhere::IndexType> no_raw_data_index_alias_set = {"IVFPQ", "IVFSQ"};
+
+    static std::set<knowhere::IndexType> conditional_hold_raw_data_index_set = {
+        IndexEnum::INDEX_HNSW,           IndexEnum::INDEX_DISKANN,     IndexEnum::INDEX_FAISS_SCANN,
+        IndexEnum::INDEX_FAISS_IVFSQ_CC, IndexEnum::INDEX_FAISS_IDMAP,
+    };
+
+    if (has_raw_data_index_set.find(indexType) != has_raw_data_index_set.end() ||
+        has_raw_data_index_alias_set.find(indexType) != has_raw_data_index_alias_set.end()) {
+        return true;
+    }
+
+    if (no_raw_data_index_set.find(indexType) != no_raw_data_index_set.end() ||
+        no_raw_data_index_alias_set.find(indexType) != no_raw_data_index_alias_set.end()) {
+        return false;
+    }
+
+    if (conditional_hold_raw_data_index_set.find(indexType) != conditional_hold_raw_data_index_set.end()) {
+        if (indexType == IndexEnum::INDEX_HNSW) {
+#ifdef KNOWHERE_WITH_CARDINAL
+            return IndexHoldRawData<DataType>(indexType, metricType, version, params);
+#else
+            return true;
+#endif
+        } else if (indexType == IndexEnum::INDEX_DISKANN) {
+#ifdef KNOWHERE_WITH_CARDINAL
+            return IndexHoldRawData<DataType>(indexType, metricType, version, params);
+#else
+            return IsMetricType(metricType, metric::L2) || IsMetricType(metricType, metric::COSINE);
+#endif
+        } else if (indexType == IndexEnum::INDEX_FAISS_SCANN) {
+            return !CheckBooleanInJson(params, indexparam::WITH_RAW_DATA, false);
+            // INDEX_FAISS_IVFSQ_CC is not online yet
+        } else if (indexType == IndexEnum::INDEX_FAISS_IVFSQ_CC) {
+            return params.find(indexparam::RAW_DATA_STORE_PREFIX) != params.end();
+        } else if (indexType == IndexEnum::INDEX_FAISS_IDMAP) {
+            if (knowhere::Version(version) <= Version::GetMinimalVersion()) {
+                return !IsMetricType(metricType, metric::COSINE);
```

Comment:
why does not cosine have a raw data? O_o

Reply:
There are some reasons, mainly because at one point, cosine was directly normalizing the raw data, which resulted in the raw data being unavailable.